syncErrorHandler - provide better details in SyncError

I'm using the `flutter_flexible_sync` example as the basis of a POC I'm working on, and after adding sync capabilities with Atlas, I'm not able to make the app work in offline mode.

Am I doing something wrong?

This is the code I'm using to create the `Realm` object.
``` dart
Future<Realm> createRealm(String appId) async {
  final user = await App(AppConfiguration(appId)).logIn(
    Credentials.anonymous(),
  );

  final realm = Realm(
    Configuration.flexibleSync(
      user,
      [
        // ...
      ],
    ),
  );
  print("Created local realm db at: ${realm.config.path}");

  realm.subscriptions.update((mutableSubscriptions) {
    mutableSubscriptions.add(realm.all<Car>());
    mutableSubscriptions.add(realm.all<Person>());
  });
  await realm.subscriptions.waitForSynchronization();
  print("Syncronization completed for realm: ${realm.config.path}");

  return realm;
}
```

And this is the issue generated when trying to launch the app in offline mode:

```
flutter: [INFO] Realm: Connection[1]: Session[1]: client_reset_config = false, Realm exists = true, client reset = false
flutter: [ERROR] Realm: Failed to resolve 'ws.realm.mongodb.com:443': Host not found (authoritative)
```

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -15,7 +15,7 @@
 * Realm Studio: 12.0.0 or later.
 
 ### Internal
-* Using Core x.y.z.
+* Using Core 12.12.0.
 
 ## 0.8.0+rc (2022-11-14)
 

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -8,6 +8,7 @@
 ### Fixed
 * Allow backlinks between files. ([#1015](https://github.com/realm/realm-dart/issues/1015))
 * Fix issue with accessing properties after traversing a backlink. ([#1018](https://github.com/realm/realm-dart/issues/1018))
+* Support mapping into `SyncSessionErrorCode` for "Compensating write" with error code 231. ([#1022](https://github.com/realm/realm-dart/pull/1022))
 
 ### Compatibility
 * Realm Studio: 12.0.0 or later.

--- a/common/lib/src/realm_types.dart
+++ b/common/lib/src/realm_types.dart
@@ -424,8 +424,13 @@ enum SyncSessionErrorCode {
   /// Client tried to open a session before initial sync is complete (BIND)
   initialSyncNotCompleted(229),
 
-  /// Client attempted a write that is disallowed by permissions, or modifies an object outside the current query - requires client reset (UPLOAD)
-  writeNotAllowed(230);
+  /// Client attempted a write that is disallowed by permissions, or modifies an object
+  /// outside the current query - requires client reset (UPLOAD)
+  writeNotAllowed(230),
+
+  /// Client attempted a write that is disallowed by permissions, or modifies an object
+  /// outside the current query, and the server undid the modificatio (UPLOAD)
+  compensatingWrite(231);
 
   static final Map<int, SyncSessionErrorCode> _valuesMap = {for (var value in SyncSessionErrorCode.values) value.code: value};
 

--- a/common/lib/src/realm_types.dart
+++ b/common/lib/src/realm_types.dart
@@ -429,7 +429,7 @@ enum SyncSessionErrorCode {
   writeNotAllowed(230),
 
   /// Client attempted a write that is disallowed by permissions, or modifies an object
-  /// outside the current query, and the server undid the modificatio (UPLOAD)
+  /// outside the current query, and the server undid the modification (UPLOAD)
   compensatingWrite(231);
 
   static final Map<int, SyncSessionErrorCode> _valuesMap = {for (var value in SyncSessionErrorCode.values) value.code: value};

--- a/test/subscription_test.dart
+++ b/test/subscription_test.dart
@@ -621,7 +621,7 @@ Future<void> main([List<String>? args]) async {
       realm.subscriptions.update((mutableSubscriptions) => mutableSubscriptions.add(query));
     }
     await realm.subscriptions.waitForSynchronization();
-    realm.write(() => realm.add(Product(ObjectId(), "doen't match subscription")));
+    realm.write(() => realm.add(Product(ObjectId(), "doesn't match subscription")));
     await realm.syncSession.waitForUpload();
 
     expect(compensatingWriteError, isA<SyncSessionError>());

--- a/test/subscription_test.dart
+++ b/test/subscription_test.dart
@@ -27,6 +27,7 @@ import '../lib/realm.dart';
 import '../lib/src/configuration.dart';
 import '../lib/src/native/realm_core.dart';
 import '../lib/src/subscription.dart';
+import 'client_reset_test.dart';
 import 'test.dart';
 
 @isTest
@@ -604,5 +605,30 @@ Future<void> main([List<String>? args]) async {
 
     realm.close();
     expect(() => subscriptions.state, throws<RealmClosedError>());
+  });
+
+  baasTest('SyncSessionErrorCode.compensatingWrite', (configuration) async {
+    late SyncError compensatingWriteError;
+    final productNamePrefix = generateRandomString(4);
+    final app = App(configuration);
+    final user = await getIntegrationUser(app);
+    final config = Configuration.flexibleSync(user, [Product.schema], syncErrorHandler: (syncError) {
+      compensatingWriteError = syncError;
+    });
+    final realm = getRealm(config);
+    final query = realm.query<Product>(r'stringQueryField BEGINSWITH $0', [productNamePrefix]);
+    if (realm.subscriptions.find(query) == null) {
+      realm.subscriptions.update((mutableSubscriptions) => mutableSubscriptions.add(query));
+    }
+    await realm.subscriptions.waitForSynchronization();
+    realm.write(() => realm.add(Product(ObjectId(), "doen't match subscription")));
+    await realm.syncSession.waitForUpload();
+
+    expect(compensatingWriteError, isA<SyncSessionError>());
+    final sessionError = compensatingWriteError.as<SyncSessionError>();
+    expect(sessionError.category, SyncErrorCategory.session);
+    expect(sessionError.isFatal, false);
+    expect(sessionError.code, SyncSessionErrorCode.compensatingWrite);
+    expect(sessionError.message!.startsWith('Client attempted a write that is outside of permissions or query filters'), isTrue);
   });
 }

